apps/dashboard/app/models/script.rb

```diff
@@ -9,8 +9,10 @@ class ClusterNotFound < StandardError; end
 
   class << self
     def scripts_dir(project_dir)
-      Pathname.new("#{project_dir}/.ondemand/scripts").tap do |path|
-        path.mkpath unless path.exist?
+      if File.dirname(project_dir).to_s == Project.dataroot.to_s
```

Comment:
When script initializes it calls `build_smart_attributes`, this in turn calls `cached_values` which 
then calls `Script.scripts_dir("#{project_dir}"), "#{id}_opts.json"` when setting the `cache_file_path`.

The root cause seemed like it is the cache relying on that `Script::sripts_dir` to check the scripts directory. I figured the fix that made the most sense and was easiest was for the `Script::scripts_dir` to check that is is in fact inside a `project_dir`.

Another way would be to leave the `Script::scripts_dir` method alone and have `cache_values` check that it is inside a project, but that seems very awkward to me.

The problem ultimately seems like a misuse by the `cached_values` with `Script.scripts_dir` which _makes_ a scripts directory, the cache seems to be using it like it would return the scripts directory. 

Should there be a separate singleton method to check this instead or to call for the path only, without the creation side effect?

Reply:
> The problem ultimately seems like a misuse by the cached_values with Script.scripts_dir which makes a scripts directory, the cache seems to be using it like it would return the scripts directory.

> Should there be a separate singleton method to check this instead or to call for the path only, without the creation side effect?

Thanks for the explanation regarding the stack of when this occurs.  It seems to me that actually Script stuff may be OK because `Script.scripts_dir` is based off of the `project_dir` which should be valid at this point (when you're making scripts).

I believe that we're actually just initializing the script object incorrectly. It's like a `garbage in -> garbage out` situation. It's doing the wrong thing because we're telling it to do the wrong thing. Here we're giving the `project_dir` a relative path like `2` (just the id) instead of the full path.

https://github.com/OSC/ondemand/blob/03156c94966fd69c75da76021ee23b8c5150bdc9/apps/dashboard/app/controllers/scripts_controller.rb#L16


Seems like this should be something more like this because at this point, the project is real with a real valid directory.

```ruby
project = Project.find(show_script_params[:project_id])
@script = Script.new(project_dir: project.directory) 
```

Reply:
I'll play with this. Seems like a much better solution since now `Script` isn't getting knowledge from the `Project` to work.